src/main/java/org/sonar/plugins/checkstyle/metadata/CheckstyleMetadata.java

```diff
@@ -0,0 +1,261 @@
+        final ClassLoader loader = getClass().getClassLoader();
+        try {
+            return Class.forName(checkName, true, loader);
+        }
+        catch (ClassNotFoundException ex) {
+            throw new IllegalStateException("exception occured during getClass for " + checkName,
+                    ex);
+        }
+    }
+
+    private boolean isTemplateRule(String checkName) {
+        return Arrays.stream(getClass(checkName).getMethods())
+                .anyMatch(CheckstyleMetadata::isSetter);
+    }
+
+    private List<String> getEnumValues(String checkName) {
+        final Class<?> loadedClass = getClass(checkName);
+        final Object[] vals = loadedClass.getEnumConstants();
+        final List<String> enumVals = new ArrayList<>();
+        for (Object val : vals) {
+            enumVals.add(val.toString());
+        }
+        return enumVals;
+    }
+
+    private void constructParams(String checkName, RulesDefinition.NewParam param,
+                                        ModulePropertyDetails modulePropertyDetails) {
+        param.setDescription(modulePropertyDetails.getDescription())
+                .setDefaultValue(modulePropertyDetails.getDefaultValue());
+        final String paramType = modulePropertyDetails.getType();
+        if ("tokens".equals(paramType) || "javadocTokens".equals(paramType)) {
+            final Object[] valuesArray = CheckUtil.getAcceptableTokens(checkName).split(",");
+            param.setType(RuleParamType.multipleListOfValues(Arrays.copyOf(
+                    valuesArray, valuesArray.length, String[].class)));
+        }
+        else if (paramType.endsWith("Option")) {
+            final Object[] valuesArray = getEnumValues(paramType).toArray();
+            param.setType(RuleParamType.singleListOfValues(Arrays.copyOf(
+                    valuesArray, valuesArray.length, String[].class)));
+        }
+        else {
+            param.setType(RuleParamType.STRING);
+        }
+    }
+
+    private static String getRuleTag(String checkPackage,
+                              SonarRulePropertyLoader.AdditionalRuleProperties configData) {
+        String result = null;
+        if (configData == null || configData.getTag() == null) {
+            final String[] packageTokens = checkPackage.split("\\.");
+            if ("puppycrawl".equals(packageTokens[1])) {
+                final String temp = packageTokens[packageTokens.length - 2];
+                if ("checks".equals(temp)) {
+                    result = "misc";
+                }
+                else {
+                    result = temp;
+                }
+            }
+        }
+        else {
+            result = configData.getTag();
+        }
+        return result;
+    }
+
+    private static Map<String, SonarRulePropertyLoader.AdditionalRuleProperties>
+        getAdditionalDetails(String fileName) {
+        final ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
```

Comment:
please add empty line above this line, to claerly show where method body is started.